### DISCUSSION TITLE
Creating exe with branch of module?

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [pyinstaller/pyinstaller#7393] [Help]
Hello - i try to bundle the following code to an exe-file on windows:

```
from datetime import datetime, timedelta
import time
import os
import sys

path = os.path.abspath(os.path.dirname(sys.argv[0]))
pathYF = os.path.join(path, "yfinance")
print(pathYF)
import sys ; sys.path.insert(0, pathYF)
import yfinance as yf

tday = datetime.today()    
startDay = tday - timedelta(days=600)          
dfPrice = yf.download("AAPL", start=startDay, end=tday)   
print(dfPrice)
time.sleep(10)
```

As you can see i installed the branch from the yfinance-tool and now the branc-module is used in the program.

No i tried to bundle this to an executable file with the following command:

`pyinstaller --onefile --collect-data G:\DEV\Fiverr\ORDER\ericromelczyk\yfinance test.py`

(i thought using this --coolect-data will collect the necessary data from the yfinance-branch to build this exe)

But now when i run the exe-file i get this error:

```
G:\DEV\Fiverr\ORDER\ericromelczyk\yfinance
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    dfPrice = yf.download("AAPL", start=startDay, end=tday)
AttributeError: module 'yfinance' has no attribute 'download'
[33472] Failed to execute script 'test' due to unhandled exception!
```

Why is this not working?
Or how can i bundle this branch into the exe-file?

**Comment by bwoodsend** [comment]
`yfinance` doesn't contain any data files so `--collect-data=yfinance` (which is the correct syntax for `--collect-data`), isn't going to do anything. Squirting random, non-existent directories into `sys.path` isn't going to do you any favours either. What error do you get without those?

**Comment by rokm** [comment]
So based on your `sys.path` modifications, your layout is as follows:
- test.py
- yfinance (the git repository checkout):
   - yfinance (the actual module within the git checkout)

When you run pyinstaller, the module search path is extended with the directory that contains the entry-point script (`test.py`). Therefore, the top-level `yfinance` directory, which is next to the entry-point script, is treated as a namespace package and collected as such (and it ends up collected as an empty namespace module, because the actual content is in yfinance sub-directory and therefore invisible to PyInstaller's import analysis). Therefore the `AttributeError: module 'yfinance' has no attribute 'download'` error - the module does exist in PyInstaller's PYZ archive, but it is an empty namespace module.

It is unclear what you are trying to accomplish here, really. 

If you just want to use the specific branch of `yfinance` in your frozen program, then it's best if you just install it in your python (virtual) environment, like you would any other package. Then there's no need to mess with `sys.path` at runtime, and the package will be automatically found and collected by PyInstaller. Or, move the actual `yfinance` package directory next to your entry-point script.

If you want to have your application dynamically load the `yfinance` from external location, then you first need to ensure that this package is not collected into the frozen application (`--exclude yfinance`). Because PYZ-collected modules and packages have precedence when loading modules in a frozen application, partially-collected package will prevent the package with same name to be discovered in external location. Then the external package (assuming it is located in the right place w.r.t. the executable) should be loaded. But I really don't recommend this approach, because PyInstaller will not analyze such a package, and you will need to manually ensure that all its dependencies (and their dependencies...) are collected.

**Reply by Rapid1898-code** [reply]
Hello - thanks for the detailed response - 
I want to simply use the yfinance branch in the created executeable

The discussion regarding this is from here:
https://github.com/ranaroussi/yfinance/issues/1321

How can i pip install the branch of yfinance using pip install?
(the name of the branch is "hotfix/sql-exception")

Is was only referred to this:
https://github.com/ranaroussi/yfinance/discussions/1080
(and this seems to be a little bit misleading)

**Reply by rokm** [reply]
`pip install git+https://github.com/ranaroussi/yfinance.git@hotfix/sql-exception`

**Reply by Rapid1898-code** [reply]
Hello - i installed with this command - but can i some how check if i now really use the branch for this module in my venv-environement?

When i do a pip list it get this info for th module:
yfinance                  0.2.4

Shouldn´t there be some other output - something with "branch"...?

**Reply by rokm** [reply]
You could just check if the code in the installed package contains the fixes you expect it to contain?

**Reply by rokm** [reply]
And if it doesn't, explicitly uninstall `yfinance` first, and then re-install using the afore-mentioned pip command (in case package is not updated/overwritten due to matching version).

**Reply by Rapid1898-code** [reply]
Thanks again a lot for your help - it worked now fine and i was able to create the exe
